### PR TITLE
feat: localized pronunciations

### PR DESCRIPTION
**Pull request by housed**
# Localized Pronunciations

Introducing the ability for developers to define locale-specific pronunciations for their skill.

## Description

If you plan to publish your skill to multiple locales, you can now define locale-specific pronunciations for your voiced responses. Simply add your locale-specific pronunciation definitions to the Litexa files that reside in said locale's Litexa project directory.

## Motivation and Context

As a developer with a multi-locale skill, you'll want to be able to define pronunciations for each locale, and you also wouldn't want pronunciations defined in one locale to carry over to another (since each locale is very likely to differ greatly in terms of pronunciations and phoneme definitions).

## Testing

Wrote Litexa integration tests that reside in `/tests/data/pronounce/` directory.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly
- [x] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

[issues]: https://github.com/alexa-games/litexa/issues
[license]: https://www.apache.org/licenses/LICENSE-2.0
[cla]: http://en.wikipedia.org/wiki/Contributor_License_Agreement

### CODE REVIEW

File: docs/reference/README.md [hunk 1]
```diff
@@ -926,6 +926,24 @@ pronounce "tomato" as "<phoneme alphabet="ipa" ph="/təˈmɑːtoʊ/">tomato</pho
...
+### localized pronunciations
+
+If you plan to publish your skill to multiple locales, you can define
+locale-specific pronunciations for your voiced responses. Simply add
+your locale-specific pronunciation definitions to the Litexa files that
+reside in said locale's Litexa project directory.
+
+::: tip Use a pronounce.litexa file in each locale's Litexa project directory
+If defining locale-specific pronunciations, we recommend creating a
+pronounce.litexa file in each of your locale's Litexa project directories.
```
Comment (ayrc) [inline docs/reference/README.md]: Add backticks to `pronounce.litexa`. If you choose to reword the title, update this description also to reflect the more general statement about file organization.

File: docs/reference/README.md [hunk 2]
```diff
@@ -926,6 +926,24 @@ pronounce "tomato" as "<phoneme alphabet="ipa" ph="/təˈmɑːtoʊ/">tomato</pho
...
+Put all of your locale-specific pronunciation definitions inside of it.
+:::
+
+::: warning The default locale's pronunciations do not carry over to other locales
```
Comment (ayrc) [inline docs/reference/README.md]: nitpick: would reword because the description is almost identical to the title. Suggestion: "Pronunciation definitions do not cascade to other locales"

File: docs/reference/README.md [hunk 3]
```diff
@@ -926,6 +926,24 @@ pronounce "tomato" as "<phoneme alphabet="ipa" ph="/təˈmɑːtoʊ/">tomato</pho
...
+Because pronunciations are very locale-specific, the default locale's pronunciations
```
Comment (ayrc) [inline docs/reference/README.md]: would put `default` in backticks to make it more noticeable, and caveat this sentence with "Unlike the structure-based override features of localization, ..." since this is the comparison we're making.

File: docs/reference/README.md [hunk 4]
```diff
@@ -926,6 +926,24 @@ pronounce "tomato" as "<phoneme alphabet="ipa" ph="/təˈmɑːtoʊ/">tomato</pho
 Note that you must use SSML in the replacement text; the Litexa
 SSML shorthand statements (e.g "<!Bravo>") will not work.
 
+### localized pronunciations
```
[the remark this answers is not in the document]
Comment (housed) [reply docs/reference/README.md]: I can do that, but was following the pattern of the other subtitles in the language reference, "slot variables" and "general use variables". I can uppercase mine and those in the next revision if we want to uppercase.

[the remark this answers is not in the document]
Comment (ayrc) [reply docs/reference/README.md]: Oops, only looked at one example. I guess, can you make them all consistent in this PR?

File: packages/litexa/src/parser/skill.coffee [hunk 5]
```diff
@@ -272,11 +272,12 @@ class lib.Skill
...
-      # TODO: support localized mappings
-      console.error "duplicate pronounciation mapping for #{source} as #{target}, previously #{@sayMapping[source]}"
-      #throw new ParserError location, "duplicate pronounciation mapping for #{source} as #{target}, previously #{@sayMapping[source]}"
-    @sayMapping[source] = target
+    for mapping in @sayMapping
+      if (mapping.language is location.language) and (mapping.source is source) and (mapping.target is not target)
+        throw new ParserError location, "duplicate pronunciation mapping for #{source} as #{target}
+          in #{location.language} language, previously #{mapping.target}"
+
+    @sayMapping.push({ language: location.language, source, target })
```
Comment (joseifert) [inline packages/litexa/src/parser/skill.coffee]: Rather than this structure:
```
[
  {
    language: 'default',
    source: 'dog',
    target: 'gerbil'
  },
  {
    language: 'default',
    source: 'cat',
    target: 'fur ball'
  }
]
```
would consider using this structure (removes the redundancy of including the language in every mapping, and speeds up iterating/filtering a language):
```
{
  'default': [
    { source: 'dog', target: 'gerbil' },
    { source: 'cat', target: 'fur ball' }
  ]
}
```
Could optionally also consider using `from/to` instead of `source/target` - think that's a little more legible in context of a mapping.

File: packages/litexa/src/parser/litexa.pegjs [hunk 6]
```diff
@@ -916,6 +916,24 @@ pronounce "tomato" as "<phoneme alphabet="ipa" ph="/təˈmɑːtoʊ/">tomato</pho
...
+If you plan to publish your skill to multiple locales, you can define
+locale-specific pronunciations for your voiced responses. Simply add
+your locale-specific pronunciation definitions to the Litexa files that
+reside in said locale's Litexa project directory.
+
+::: tip Use a pronounce.litexa file in each locale's Litexa project directory
+If defining locale-specific pronunciations, we recommend creating a
+pronounce.litexa file in each of your locale's Litexa project directories.
+Put all of your locale-specific pronunciation definitions inside of it.
+:::
```
Comment (joseifert) [inline packages/litexa/src/parser/litexa.pegjs]: Would include the below tip in this paragraph and slightly abbreviate it (since it's somewhat contingent on project setup). Perhaps something like: ... in said locale's Litexa project directory (see the Localization chapter, for further information). We recommend using a standalone file (e.g. `pronounce.litexa`) for pronunciation definitions.

File: tests/data/pronounce/litexa/main.litexa [hunk 7]
```diff
@@ -25,3 +21,48 @@ TEST "say no"
   alexa: launch, /I really like my cat/i
   user: AMAZON.NoIntent
   alexa: null, /goodbye then/i
+
+TEST "say yes en-CA"
```
Comment (joseifert) [inline tests/data/pronounce/litexa/main.litexa]: Could you update these tests to also regex check that the phoneme pronunciation is being applied to all languages except for en-IN?

File: packages/litexa/src/parser/skill.coffee [hunk 8]
```diff
@@ -363,11 +364,12 @@ class lib.Skill
...
-        source = source.replace(/'/g, '\\\'')
-        target = target.replace(/'/g, '\\\'')
-        lines.push "  { test: new RegExp(' #{source}','gi'), change: ' #{target}' }"
-        lines.push "  { test: new RegExp('#{source} ','gi'), change: '#{target} ' }"
+      for mapping in @sayMapping
+        source = mapping.source.replace /'/g, '\\\''
+        target = mapping.target.replace /'/g, '\\\''
+        language = mapping.language
+        lines.push "  { test: new RegExp(' #{source}','gi'), change: ' #{target}', language: '#{language}' }"
+        lines.push "  { test: new RegExp('#{source} ','gi'), change: '#{target} ', language: '#{language}' }"
```
Comment (joseifert) [inline packages/litexa/src/parser/skill.coffee]: Odd to have the naming suddenly flipped to `test/change` here. Would align the naming with the function parameters. For instance, if you were to go with `from/to` you could use something like `fromRegex` and `toRegex`.

File: tests/data/pronounce/litexa/main.litexa (hunk 7 above)
Comment (housed) [reply tests/data/pronounce/litexa/main.litexa]: Could you be more specific? I do `alexa: null, /phoneme/` in all tests except for en-IN.

File: packages/litexa/src/parser/skill.coffee [hunk 9]
```diff
@@ -364,12 +367,12 @@ class lib.Skill
...
-        target = mapping.target.replace /'/g, '\\\''
-        language = mapping.language
-        lines.push "  { test: new RegExp(' #{source}','gi'), change: ' #{target}', language: '#{language}' }"
-        lines.push "  { test: new RegExp('#{source} ','gi'), change: '#{target} ', language: '#{language}' }"
+      for language of @sayMapping
+        for mapping in @sayMapping[language]
+          from = mapping.from.replace /'/g, '\\\''
+          to = mapping.to.replace /'/g, '\\\''
+          lines.push "  { from: new RegExp(' #{from}','gi'), to: ' #{to}', language: '#{language}' }"
+          lines.push "  { from: new RegExp('#{from} ','gi'), to: '#{to} ', language: '#{language}' }"
```
Comment (joseifert) [inline packages/litexa/src/parser/skill.coffee]: Don't need the language in the mapping any longer, since it's now the parent.

File: tests/data/pronounce/litexa/main.litexa [hunk 10]
```diff
@@ -11,39 +11,39 @@ launch
...
 
 TEST "say yes en-CA"
   setRegion en-CA
   launch
-  alexa: launch, /I really like my ferret/i
+  alexa: launch, /I really like my ferret/
   user: AMAZON.YesIntent
   alexa: null, /my ferret/
-  alexa: null, /phoneme/
+  alexa: null, /phoneme.*concealed/i
```
Comment (joseifert) [inline tests/data/pronounce/litexa/main.litexa]: Could combine these into the same regex check. Something like: `/my ferret.*<phoneme[^>]*>concealed`/

File: packages/litexa/src/parser/litexa.pegjs [hunk 11]
```diff
@@ -917,22 +917,22 @@ pronounce "tomato" as "<phoneme alphabet="ipa" ph="/təˈmɑːtoʊ/">tomato</pho
...
 your locale-specific pronunciation definitions to the Litexa files that
-reside in said locale's Litexa project directory.
+reside in said locale's Litexa project directory (see the Localization
+chapter for further information).
 
-::: tip Use a pronounce.litexa file in each locale's Litexa project directory
-If defining locale-specific pronunciations, we recommend creating a
-pronounce.litexa file in each of your locale's Litexa project directories.
-Put all of your locale-specific pronunciation definitions inside of it.
+::: tip Tip
```
[the remark this answers is not in the document]
Comment (housed) [reply packages/litexa/src/parser/litexa.pegjs]: Good to know about auto-header. But just as well, searching through the codebase, there seems to be inconsistent uses:

1. Currently see a lot of both `:::tip` and `::: tip`.
2. Auto-header is all uppercase, but most of our manual uses are regular-case headings.

I'll remove the space, but keep the manual heading so that it is regular casing.

[the remark this answers is not in the document]
Comment (joseifert) [reply packages/litexa/src/parser/litexa.pegjs]: Ah, right - forgot the auto-header is fully upper case.

If the spacing after the `:::` is inconsistent anyhow, go ahead and disregard this comment and leave as is.
